bolt.rb

```diff
@@ -117,36 +117,55 @@ class Target < Wash::Entry
     }
   end
 
-  def known_hosts(host_key_check)
-    return nil unless host_key_check == false
-
-    # Disable host key checking by redirecting known hosts to an empty file
-    # This is future-proofing for when Wash works on Windows.
-    Gem.win_platform? ? 'NUL' : '/dev/null'
-  end
-
-  def transport_options(target)
-    {
-      host: target.host,
-      port: target.port,
-      user: target.user,
-      password: target.password,
-      identity_file: target.options['private-key'],
-      known_hosts: known_hosts(target.options['host-key-check'])
-    }
-  end
-
   def initialize(target)
     # Save just the target information we need as state.
     @name = target.name
     @partial_metadata = target.detail
-    # TODO: add WinRM
-    transport :ssh, transport_options(target) if target.protocol == 'ssh'
+    @target = target.to_h
     prefetch :list
   end
 
-  def exec(*_args)
-    raise 'non-ssh protocols are not yet implemented'
+  # Only implements SSH and WinRM. Docker can use the Wash Docker plugin, local
+  # is trivial, and remote is not really usable. PCP I hope to implement later.
+  def exec(cmd, args, opts)
+    # lazy-load dependencies to make the plugin as fast as possible
+    require 'bolt/target'
+    require 'logging'
+    require 'shellwords'
+
+    # opts can contain 'tty', 'stdin', and 'elevate'. If tty is set, apply it
+    # to the target for this exec.
+    target_opts = @target.transform_keys(&:to_s)
+    target_opts['tty'] = true if opts[:tty]
+    target = Bolt::Target.new(@target[:uri], target_opts)
+
+    logger = Logging.logger($stderr)
+    logger.level = :warn
+    command = Shellwords.join([cmd] + args)
+
+    transport = target.transport || 'ssh'
+    case transport
+    when 'ssh'
+      require_relative 'transport_ssh.rb'
+      connection = BoltSSH.new(target, logger)
+    when 'winrm'
+      require_relative 'transport_winrm.rb'
+      connection = BoltWinRM.new(target, logger)
+    else
+      raise "#{transport} unsupported"
+    end
+
+    begin
+      connection.connect
+      # Returns exit code
+      connection.execute(command, stdin: opts[:stdin] ? $stdin : nil)
+    ensure
+      begin
+        connection&.disconnect
```

Comment:
Should this be in a SIGTERM handler as well?

Reply:
Ensure handles that.